File: 1694-make-sum-divisible-by-p/make-sum-divisible-by-p.py
class Solution:
    def minSubarray(self, nums: List[int], p: int) -> int:
        n=len(nums)
        parray=[0]*(n+1)
        for i in range(n):
            parray[i+1]=parray[i]+nums[i]
        total=parray[n]
        if total%p==0:
            return 0
        mapp={}
        maxx=0
        for i in range(n):
            second=(total-parray[i])%p
            first=(total-second)%p
            slen=n-(i)
            flen=n-slen
            temp=p-second
            print(i)
            print(second,first)
            print(slen,flen)
            if first==0:
                maxx=max(maxx,flen)
            if second==0 and 0 in mapp:
                maxx=max(maxx,mapp[0]+slen)
            if temp in mapp:
                maxx=max(maxx,mapp[temp]+slen)
    
            if first in mapp:
                mapp[first]=max(mapp[first],flen)
            else:
                mapp[first]=flen

        return n-maxx if maxx!=0 else -1



        

File: 1694-make-sum-divisible-by-p/make-sum-divisible-by-p.java
class Solution {
    public int minSubarray(int[] nums, int p) {
        int n = nums.length;
        int[] parray = new int[n + 1];

        // Prefix sum modulo p
        for (int i = 0; i < n; i++) {
            parray[i + 1] = (parray[i] + nums[i]) % p;
        }

        int total = parray[n];
        if (total % p == 0) return 0;

        HashMap<Integer, Integer> map = new HashMap<>();
        int maxx = 0;

        for (int i = 0; i < n; i++) {
            int second = ((total - parray[i]) % p + p) % p;
            int first = ((total - second) % p + p) % p;

            int slen = n - i;
            int flen = n - slen;

            int temp = (p - second) % p;

            if (first == 0) {
                maxx = Math.max(maxx, flen);
            }

            if (second == 0 && map.containsKey(0)) {
                maxx = Math.max(maxx, map.get(0) + slen);
            }

            if (map.containsKey(temp)) {
                maxx = Math.max(maxx, map.get(temp) + slen);
            }

            if (map.containsKey(first)) {
                map.put(first, Math.max(map.get(first), flen));
            } else {
                map.put(first, flen);
            }
        }

        return maxx != 0 ? n - maxx : -1;
    }
}
